contracts/interchain-token-service.clar

```diff
@@ -0,0 +1,509 @@
+(use-trait native-interchain-token-trait .traits.native-interchain-token-trait)
+(use-trait gateway-trait .traits.gateway-trait)
+
+;; token definitions
+;;
+
+;; constants
+;;
+(define-constant ERR-NOT-AUTHORIZED (err u21051))
+
+(define-constant DEPLOYER tx-sender)
+
+(define-public (set-paused (its-impl <its-trait>) (status bool))
+    (begin
+        (asserts! (is-correct-impl its-impl) ERR-INVALID-IMPL)
+        (contract-call? its-impl set-paused status contract-caller)))
+
+
+;; ####################
+;; ####################
+;; ### Operatorship ###
+;; ####################
+;; ####################
+
+;; Transfers operatorship to a new account
+(define-public (transfer-operatorship (its-impl <its-trait>) (new-operator principal))
+    (begin
+        (asserts! (is-correct-impl its-impl) ERR-INVALID-IMPL)
+        (contract-call? its-impl transfer-operatorship new-operator contract-caller)
+    )
+)
+
+(define-public (transfer-ownership (its-impl <its-trait>) (new-owner principal))
+    (begin
+        (asserts! (is-correct-impl its-impl) ERR-INVALID-IMPL)
+        (contract-call? its-impl transfer-ownership new-owner contract-caller)
+    )
+)
+
+;; ####################
+;; ####################
+;; ### address tracking ###
+;; ####################
+;; ####################
+
+;; Sets the trusted address and its hash for a remote chain
+;; @param its-impl The implementation of the Interchain Token Service
+;; @param chain-name Chain name of the remote chain
+;; @param address the string representation of the trusted address
+;; #[allow(unchecked_data)]
+(define-public (set-trusted-address (its-impl <its-trait>) (chain-name (string-ascii 19)) (address (string-ascii 128)))
+    (begin
+        (asserts! (is-correct-impl its-impl) ERR-INVALID-IMPL)
+        (contract-call? its-impl set-trusted-address chain-name address contract-caller)))
+
+;; Remove the trusted address of the chain.
+;; @param its-impl The implementation of the Interchain Token Service
+;; @param chain-name Chain name that should be made untrusted
+;; #[allow(unchecked_data)]
+(define-public (remove-trusted-address (its-impl <its-trait>) (chain-name  (string-ascii 19)))
+    (begin
+        (asserts! (is-correct-impl its-impl) ERR-INVALID-IMPL)
+        (contract-call? its-impl remove-trusted-address chain-name contract-caller)))
+
+(define-read-only (get-its-hub-chain)
+    (contract-call? .interchain-token-service-storage get-its-hub-chain))
+(define-read-only (get-trusted-address (chain (string-ascii 19)))
+    (contract-call? .interchain-token-service-storage get-trusted-address chain))
+
+
+(define-private (get-call-params (destination-chain (string-ascii 19)) (payload (buff 63000)))
+    (let (
+            (destination-address (unwrap! (get-trusted-address destination-chain) ERR-UNTRUSTED-CHAIN))
+            (destination-address-hash (keccak256 (unwrap-panic (to-consensus-buff? destination-address))))
```

Comment:
what is this for? it appears these variables are not used? also, i don't know that the trusted address of the destination chain will be set, since all transfers just go through the hub, and the hub tracks the trusted address for each chain.